lib/new_relic/traced_thread.rb

```diff
@@ -0,0 +1,27 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+module NewRelic
+  class TracedThread < Thread
+    def initialize(*args, &block)
+      # we should add metrics
+      traced_block = create_traced_block(*args, block)
+      super(*args, &traced_block)
+    end
+
+    def create_traced_block(*args, block)
+      return block if NewRelic::Agent.config[:'instrumentation.thread.tracing'] # if this is on, don't double trace
+
+      instrumentation = ::Thread.current[:newrelic_tracer_state]
+      Proc.new do |*args|
+        ::Thread.current[:newrelic_tracer_state] = instrumentation
```

Comment:
I'd love a demo / overview of this bit when you get a chance. I'm not understanding how the tracer state gets set originally in order for it to be passed as `instrumentation` to the Proc. From looking at the tests, it seems like they explicitly set the state beforehand. Is doing so required?

Reply:
Sure, i'd be happy to go over this stuff! How this works was mostly taken from ancient lore uncovered on google at one point about how to make threads work in the agent (aka some super old github gist from like 8 or 10 years ago irrc, i'll see if i can find it again)

As for where it gets set originally, there is the method `state_for` in `lib/new_relic/agent/transaction.rb` that will create a new state and set it if it doesn't exist yet.

Reply:
Oh not quite as old as I thought it was https://gist.github.com/mikeantonelli/0d7563db4ad28d7ee10d365b295ce3f4